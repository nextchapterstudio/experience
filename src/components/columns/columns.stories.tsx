import React from 'react'
import { ComponentStory, ComponentMeta } from '@storybook/react'

import { Column, Columns, FullColumn } from './columns'
import { Content } from '../content'

// More on default export: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
export default {
  title: 'Example/Columns',
  component: Columns,
} as ComponentMeta<typeof Columns>

// More on component templates: https://storybook.js.org/docs/react/writing-stories/introduction#using-args
const Template: ComponentStory<typeof Column> = (args) => (
  <Columns>
    <Column {...args}>One Half</Column>
    <Column {...args}>One Half</Column>
  </Columns>
)

export const IsHalf = Template.bind({})
// More on args: https://storybook.js.org/docs/react/writing-stories/args
IsHalf.args = {
  isHalf: true,
}
const MultiTemplate: ComponentStory<typeof Columns> = (args) => (
  <Columns {...args}>
    <Column>One</Column>
    <Column>Two</Column>
    <Column>Three</Column>
    <Column>Four</Column>
  </Columns>
)
export const IsMulti = MultiTemplate.bind({})
// More on args: https://storybook.js.org/docs/react/writing-stories/args
IsMulti.args = {}

const SingleTemplate: ComponentStory<typeof Columns> = (args) => (
  <Columns {...args}>
    <Column isHalf>One</Column>
  </Columns>
)
export const IsSingle = SingleTemplate.bind({})
// More on args: https://storybook.js.org/docs/react/writing-stories/args
IsSingle.args = {}
