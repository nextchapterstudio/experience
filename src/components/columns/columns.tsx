import React, { ElementType, FC, ReactNode, HTMLAttributes } from 'react'
import classnames from 'classnames'

export interface ColumnsProps extends HTMLAttributes<HTMLOrSVGElement> {
  children?: React.ReactNode
  className?: string
  isMobile?: boolean
  isMultiline?: boolean
  isCentered?: boolean
  isNarrow?: boolean
  isGapless?: boolean
}
export const Columns: FC<ColumnsProps> = ({
  children,
  className,
  isMobile,
  isMultiline,
  isCentered,
  isNarrow,
  isGapless,
  ...props
}) => {
  const classes = {
    'is-mobile': isMobile,
    'is-multiline': isMultiline,
    'is-centered': isCentered,
    'is-narrow': isNarrow,
    'is-gapless': isGapless,
  }

  return (
    <div className={classnames('columns', className, classes)} {...props}>
      {children}
    </div>
  )
}
Columns.defaultProps = {
  className: '',
  isMobile: false,
  isMultiline: false,
  isCentered: false,
  isNarrow: false,
  isGapless: false,
}

export interface ColumnProps extends HTMLAttributes<HTMLOrSVGElement> {
  as?: ElementType
  children?: ReactNode
  className?: string
  isSpaced?: boolean
  is?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12' | null
  isThreeQuarters?: boolean
  isTwoThirds?: boolean
  isHalf?: boolean
  isOneThird?: boolean
  isOneQuarter?: boolean
  isFull?: boolean
  isFourFifths?: boolean
  isThreeFifths?: boolean
  isTwoFifths?: boolean
  isOneFifth?: boolean
  isNarrow?: boolean
  isOffsetOneQuarter?: boolean
  isOffsetOneFifth?: boolean
  isOffsetHalf?: boolean
  isOffsetOneThird?: boolean
  isHidden?: boolean
  isFullheight?: boolean
  offset?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | null
  size?: '1' | '2' | '3' | '4' | '5' | '6' | '7' | '8' | '9' | '10' | '11' | '12' | null
}
export const Column: FC<ColumnProps> = ({
  children,
  as: Col = 'div',
  className,
  isThreeQuarters,
  isTwoThirds,
  isHalf,
  isOneThird,
  isOneQuarter,
  isFull,
  isFourFifths,
  isThreeFifths,
  isTwoFifths,
  isOneFifth,
  isNarrow,
  isOffsetOneQuarter,
  isOffsetOneFifth,
  isOffsetHalf,
  isOffsetOneThird,
  isHidden,
  offset,
  size,
  is,
  isFullheight,
  ...props
}) => {
  const classes = {
    'is-four-fifths': isFourFifths,
    'is-three-fifths': isThreeFifths,
    'is-two-fifths': isTwoFifths,
    'is-one-fifth': isOneFifth,
    'is-three-quarters': isThreeQuarters,
    'is-two-thirds': isTwoThirds,
    'is-half': isHalf,
    'is-one-third': isOneThird,
    'is-one-quarter': isOneQuarter,
    'is-full': isFull,
    'is-narrow': isNarrow,
    'is-offset-one-quarter': isOffsetOneQuarter,
    'is-offset-one-fifth': isOffsetOneFifth,
    'is-offset-half': isOffsetHalf,
    'is-offset-one-third': isOffsetOneThird,
    'is-hidden': isHidden,
    'is-1': false,
    'is-2': false,
    'is-3': false,
    'is-4': false,
    'is-5': false,
    'is-6': false,
    'is-7': false,
    'is-8': false,
    'is-9': false,
    'is-10': false,
    'is-11': false,
    'is-12': false,
    'is-offset-1': false,
    'is-offset-2': false,
    'is-offset-3': false,
    'is-offset-4': false,
    'is-offset-5': false,
    'is-offset-6': false,
    'is-offset-7': false,
    'is-offset-8': false,
    'is-offset-9': false,
    'is-offset-10': false,
    'is-offset-11': false,
    'is-full-height': isFullheight,
  }
  if (size) {
    classes[`is-${size}`] = true
  }
  if (is) {
    classes[`is-${is}`] = true
  }
  if (offset) {
    classes[`is-offset-${offset}`] = true
  }
  if (size && is) {
    console.warn('Do not use both size and is')
  }
  return (
    <Col className={classnames('column', className, classes)} {...props}>
      {children}
    </Col>
  )
}

Column.defaultProps = {
  isFourFifths: false,
  isThreeFifths: false,
  isTwoFifths: false,
  isOneFifth: false,
  isThreeQuarters: false,
  isTwoThirds: false,
  isHalf: false,
  isOneThird: false,
  isOneQuarter: false,
  isFull: false,
  isNarrow: false,
  as: 'div',
}

export interface FullColumnProps {
  children?: React.ReactNode
  className?: string
}
export const FullColumn: FC<FullColumnProps> = ({ children, className }: FullColumnProps) => (
  <Columns>
    <Column isFull className={className}>
      {children}
    </Column>
  </Columns>
)
