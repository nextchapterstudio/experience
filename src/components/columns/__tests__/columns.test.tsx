import React from 'react'
import { render, cleanup, fireEvent, screen } from '@testing-library/react'
import { Column, Columns, FullColumn } from '../columns'

describe('The Columns Component', () => {
  afterEach(cleanup)
  it('should render correct structure ', () => {
    const { container } = render(<Columns><Column offset='4'>Offset</Column></Columns>)
    expect(container).toMatchSnapshot()
    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toHaveClass('columns')
    expect(container?.firstChild?.firstChild).toHaveClass('column')
    expect(container?.firstChild?.firstChild).toHaveClass('is-offset-4')
  })

  it('should render correct structure for the FullColumn', () => {
    const { container } = render(<FullColumn>Offset</FullColumn>)
    expect(container).toMatchSnapshot()
    expect(container.firstChild).toMatchSnapshot()
    expect(container.firstChild).toHaveClass('columns')
    expect(container?.firstChild?.firstChild).toHaveClass('column')
    expect(container?.firstChild?.firstChild).toHaveClass('is-full')
  })
})
