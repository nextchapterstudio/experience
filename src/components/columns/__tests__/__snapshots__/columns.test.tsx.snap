// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Columns Component should render correct structure  1`] = `
<div>
  <div
    class="columns"
  >
    <div
      class="column is-offset-4"
    >
      Offset
    </div>
  </div>
</div>
`;

exports[`The Columns Component should render correct structure  2`] = `
<div
  class="columns"
>
  <div
    class="column is-offset-4"
  >
    Offset
  </div>
</div>
`;

exports[`The Columns Component should render correct structure for the FullColumn 1`] = `
<div>
  <div
    class="columns"
  >
    <div
      class="column is-full"
    >
      Offset
    </div>
  </div>
</div>
`;

exports[`The Columns Component should render correct structure for the FullColumn 2`] = `
<div
  class="columns"
>
  <div
    class="column is-full"
  >
    Offset
  </div>
</div>
`;
